lms/services/course.py

```diff
@@ -1,22 +1,39 @@
 import json
 from copy import deepcopy
 
-from lms.models import CourseGroupsExportedFromH, LegacyCourse
+from lms.models import Course, CourseGroupsExportedFromH, LegacyCourse
 
 
 class CourseService:
     def __init__(self, application_instance_service, consumer_key, db):
-        self._application_instance_service = application_instance_service
+        self._application_instance = application_instance_service.get()
         self._consumer_key = consumer_key
         self._db = db
 
     def get_or_create(self, authority_provided_id):
-        """
-        Add the current course to the `course` table if it's not there already.
+        """Add the current course to the `course` table if it's not there already."""
+        return self._get_legacy(authority_provided_id) or self._create_legacy(
+            authority_provided_id
+        )
 
-        :raise ConsumerKeyError: if request.lti_user.oauth_consumer_key isn't in the DB
-        """
-        return self._get(authority_provided_id) or self._create(authority_provided_id)
+    def upsert(
+        self, authority_provided_id, context_id, name, extra, settings=None
+    ):  # pylint: disable=too-many-arguments
+        course = self._get(authority_provided_id)
+
+        if not course:
+            course = self._create(
+                authority_provided_id,
+                context_id,
+                name,
+                extra,
+                settings,
+            )
+
+        # Update any values that might have changed
+        course.lms_name = name
```

Comment:
I think this is the only thing we want to be part of the update (settings are not updated on the app as far as I can tell)

Reply:
Maybe `extra` should also be updated?